docs/man/enrich.1
.\" Automatically generated by Pandoc 3.2
.\"
.TH "TRACEE\-ENRICH" "1" "2025/04" "" "Tracee Enrich Flag Manual"
.SS NAME
tracee \f[B]\-\-enrich\f[R] \- Configure enrichment for container events
and other enrichment options
.SS SYNOPSIS
tracee \f[B]\-\-enrich\f[R]
[container.enabled=\f[I]true|false\f[R]|container.cgroup.path=\f[I]path\f[R]|container.docker.socket=\f[I]socket_path\f[R]|container.containerd.socket=\f[I]socket_path\f[R]|container.crio.socket=\f[I]socket_path\f[R]|container.podman.socket=\f[I]socket_path\f[R]|resolve\-fd=\f[I]true|false\f[R]|exec\-hash.enabled=\f[I]true|false\f[R]|exec\-hash.mode=\f[I]mode\f[R]|user\-stack\-trace=\f[I]true|false\f[R]]
[\f[B]\-\-enrich\f[R] \&...]
.SS DESCRIPTION
The \f[CR]\-\-enrich\f[R] flag allows you to configure enrichment
options for container events and other enrichment features.
.SS Flags
.IP \[bu] 2
\f[B]container.enabled\f[R]=\f[I]true|false\f[R] Enable or disable
container enrichment.
Example:
.RS 2
.IP
.EX
\-\-enrich container.enabled=true
.EE
.RE
.IP \[bu] 2
\f[B]container.cgroup.path\f[R]=\f[I]path\f[R] Configure the path to the
cgroupfs where container cgroups are created.
This is used as a hint for auto\-detection.
Example:
.RS 2
.IP
.EX
\-\-enrich container.cgroup.path=/sys/fs/cgroup
.EE
.RE
.IP \[bu] 2
\f[B]container.docker.socket\f[R]=\f[I]socket_path\f[R] Configure the
path to the docker socket.
Example:
.RS 2
.IP
.EX
\-\-enrich container.docker.socket=/var/run/docker.sock
.EE
.RE
.IP \[bu] 2
\f[B]container.containerd.socket\f[R]=\f[I]socket_path\f[R] Configure
the path to the containerd socket.
Example:
.RS 2
.IP
.EX
\-\-enrich container.containerd.socket=/var/run/containerd/containerd.sock
.EE
.RE
.IP \[bu] 2
\f[B]container.crio.socket\f[R]=\f[I]socket_path\f[R] Configure the path
to the crio socket.
Example:
.RS 2
.IP
.EX
\-\-enrich container.crio.socket=/var/run/crio/crio.sock
.EE
.RE
.IP \[bu] 2
\f[B]container.podman.socket\f[R]=\f[I]socket_path\f[R] Configure the
path to the podman socket.
Example:
.RS 2
.IP
.EX
\-\-enrich container.podman.socket=/var/run/podman/podman.sock
.EE
.RE
.IP \[bu] 2
\f[B]resolve\-fd\f[R]=\f[I]true|false\f[R] Enable or disable
resolve\-fd.
Example:
.RS 2
.IP
.EX
\-\-enrich resolve\-fd=true
.EE
.RE
.IP \[bu] 2
\f[B]exec\-hash.enabled\f[R]=\f[I]true|false\f[R] Enable or disable
exec\-hash.
Example:
.RS 2
.IP
.EX
\-\-enrich exec\-hash.enabled=true
.EE
.RE
.IP \[bu] 2
\f[B]exec\-hash.mode\f[R]=\f[I]mode\f[R] Configure the mode for
exec\-hash.
Example:
.RS 2
.IP
.EX
\-\-enrich exec\-hash.mode=sha256
.EE
.RE
.IP \[bu] 2
\f[B]user\-stack\-trace\f[R]=\f[I]true|false\f[R] Enable or disable
user\-stack\-trace.
Example:
.RS 2
.IP
.EX
\-\-enrich user\-stack\-trace=true
.EE
.RE
.SS EXAMPLES
.IP "1." 3
Enable container enrichment:
.RS 4
.IP
.EX
\-\-enrich container.enabled=true
.EE
.RE
.IP "2." 3
Configure Docker socket:
.RS 4
.IP
.EX
\-\-enrich container.docker.socket=/var/run/docker.sock
.EE
.RE
.IP "3." 3
Set the cgroupfs path:
.RS 4
.IP
.EX
\-\-enrich container.cgroup.path=/sys/fs/cgroup
.EE
.RE
.IP "4." 3
Combine multiple flags:
.RS 4
.IP
.EX
\-\-enrich container.enabled=true container.docker.socket=/var/run/docker.sock container.cgroup.path=/sys/fs/cgroup
.EE
.RE
.IP "5." 3
Enable resolve\-fd and exec\-hash:
.RS 4
.IP
.EX
\-\-enrich resolve\-fd=true exec\-hash.enabled=true
.EE
.RE
.PP
Please refer to the documentation for more information on container
events enrichment.


pkg/cmd/flags/enrich.go
package flags

import (
	"strings"

	"github.com/aquasecurity/tracee/common/errfmt"
	"github.com/aquasecurity/tracee/pkg/datastores/container/runtime"
)

const (
	EnrichFlag                    = "enrich"
	ContainerEnabledFlag          = "container.enabled"
	ContainerCgroupPathFlag       = "container.cgroup.path"
	ContainerDockerSocketFlag     = "container.docker.socket"
	ContainerContainerdSocketFlag = "container.containerd.socket"
	ContainerCrioSocketFlag       = "container.crio.socket"
	ContainerPodmanSocketFlag     = "container.podman.socket"
	ResolveFdFlag                 = "resolve-fd"
	ExecHashEnabledFlag           = "exec-hash.enabled"
	ExecHashModeFlag              = "exec-hash.mode"
	UserStackTraceFlag            = "user-stack-trace"

	enrichInvalidFlagFormat = "invalid enrichment flag: %s, use 'trace man enrichment' for more info"
)

func enrichmentHelp() string {
	return `Configure enrichment for container events.

Flags:
  --enrich container.enabled=<true/false>       Enable or disable container enrichment.
  --enrich container.cgroup.path=             Configure the path to the cgroupfs where container cgroups are created. This is used as a hint for auto-detection.
  --enrich container.docker.socket=<socket_path> Configure the path to the docker socket.
  --enrich container.containerd.socket=<socket_path> Configure the path to the containerd socket.
  --enrich container.crio.socket=<socket_path> Configure the path to the crio socket.
  --enrich container.podman.socket=<socket_path> Configure the path to the podman socket.
  --enrich resolve-fd=<true/false>             Enable or disable resolve-fd.
  --enrich exec-hash.enabled=<true/false>     Enable or disable exec-hash.
  --enrich exec-hash.mode=                    Configure the mode for exec-hash.
  --enrich user-stack-trace=<true/false>      Enable or disable user-stack-trace.
`
}

type EnrichmentConfig struct {
	ContainerEnabled          bool
	ContainerCgroupPath       string
	ContainerDockerSocket     string
	ContainerContainerdSocket string
	ContainerCrioSocket       string
	ContainerPodmanSocket     string
	ResolveFd                 bool
	ExecHashEnabled           bool
	ExecHashMode              string
	UserStackTrace            bool
}

func PrepareEnrichment(enrichment []string) (EnrichmentConfig, error) {
	var enrichmentConfig EnrichmentConfig

	for _, flag := range enrichment {
		parts := strings.Split(flag, "=")
		if len(parts) != 2 && !IsBoolFlag(parts[0]) {
			return EnrichmentConfig{}, errfmt.Errorf(enrichInvalidFlagFormat, flag)
		}

		if len(parts) > 1 && IsBoolFlag(parts[0]) {
			return EnrichmentConfig{}, errfmt.Errorf(enrichInvalidFlagFormat, flag)
		}

		flagName := parts[0]

		switch flagName {
		case ContainerEnabledFlag:
			enrichmentConfig.ContainerEnabled = true
		case ContainerCgroupPathFlag:
			enrichmentConfig.ContainerCgroupPath = parts[1]
		case ContainerDockerSocketFlag:
			enrichmentConfig.ContainerDockerSocket = parts[1]
		case ContainerContainerdSocketFlag:
			enrichmentConfig.ContainerContainerdSocket = parts[1]
		case ContainerCrioSocketFlag:
			enrichmentConfig.ContainerCrioSocket = parts[1]
		case ContainerPodmanSocketFlag:
			enrichmentConfig.ContainerPodmanSocket = parts[1]
		case ResolveFdFlag:
			enrichmentConfig.ResolveFd = true
		case ExecHashEnabledFlag:
			enrichmentConfig.ExecHashEnabled = true
		case ExecHashModeFlag:
			enrichmentConfig.ExecHashMode = parts[1]
		case UserStackTraceFlag:
			enrichmentConfig.UserStackTrace = true
		default:
			return EnrichmentConfig{}, errfmt.Errorf(enrichInvalidFlagFormat, flagName)
		}
	}

	return enrichmentConfig, nil
}

func IsBoolFlag(flag string) bool {
	return flag == "container.enabled" || flag == "resolve-fd" || flag == "exec-hash.enabled" || flag == "user-stack-trace"
}

func (e *EnrichmentConfig) GetRuntimeSockets() (runtime.Sockets, error) {
	sockets := runtime.Sockets{}
	if e.ContainerDockerSocket != "" {
		err := sockets.Register(runtime.Docker, e.ContainerDockerSocket)
		if err != nil {
			return sockets, errfmt.Errorf("failed to register docker socket: %v", err)
		}
	}
	if e.ContainerContainerdSocket != "" {
		err := sockets.Register(runtime.Containerd, e.ContainerContainerdSocket)
		if err != nil {
			return sockets, errfmt.Errorf("failed to register containerd socket: %v", err)
		}
	}
	if e.ContainerCrioSocket != "" {
		err := sockets.Register(runtime.Crio, e.ContainerCrioSocket)
		if err != nil {
			return sockets, errfmt.Errorf("failed to register crio socket: %v", err)
		}
	}
	if e.ContainerPodmanSocket != "" {
		err := sockets.Register(runtime.Podman, e.ContainerPodmanSocket)
		if err != nil {
			return sockets, errfmt.Errorf("failed to register podman socket: %v", err)
		}
	}
	return sockets, nil
}
